### DISCUSSION TITLE
Disable the `unicorn/no-array-reduce` rule

### DISCUSSION THREAD

**Discussion by leroykorterink** [mediamonks/eslint-config#229] [General]
There is a proposal to disable the `unicorn/no-array-reduce` rule.

> `unicorn/no-array-reduce`: super annoying. introduces unnecessary verbosity
> \- @edgardz

**Comment by leroykorterink** [comment]
`Array#reduce()` and `Array#reduceRight()` usually result in [hard-to-read](https://twitter.com/jaffathecake/status/1213077702300852224) and [less performant](https://www.richsnapp.com/article/2019/06-09-reduce-spread-anti-pattern) code. In almost every case, it can be replaced by `.map`, `.filter`, or a `for-of` loop.

It's only somewhat useful in the rare case of summing up numbers, which is allowed by default.
 Use `eslint-disable` comment if you really need to use `Array#reduce()`.

Please see [the documentation](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-reduce.md) of the rule.

**Reply by ThaNarie** [reply]
It would really be useful to have code examples where people really want to use `reduce` compared to suggested alternatives.

Then we can compare both variants and have a proper discussion.

**Comment by edgardz** [comment]
I use reduce all the time. It is a very powerful language feature we should not disable because some people are not familiar with it.
Here is an example I used it recently:

```
const x = 123

const children: HTMLElement[] = [...]

const closestChildIndex = children.reduce((previous, current, index) => {
  return Math.abs(current.offsetLeft + x) < Math.abs(children[previous].offsetLeft + x) ? index : previous
}, 0)
```

All the logic to achieve the closestChildIndex value is contained inside the reduce. 

Otherwise I would have to do something like:

```
const x = 123

const children: HTMLElement[] = [...]

let closestChildIndex = 0
let closestChildOffset = Math.abs(children[0].offsetLeft + x)

for (let index = 1; index < children.length; index++) {
  const currentOffset = Math.abs(children[index].offsetLeft + x)
  if (currentOffset < closestChildOffset) {
    closestChildIndex = index
    closestChildOffset = currentOffset
  }
}
```

I don't know if I'm too familiar with how reduce works, but the first option is easier to read.

**Reply by ThaNarie** [reply]
I agree that that's a perfectly valid use case for `reduce`, and the best way to write it if you indeed need the index (and not the item).
